### PR TITLE
Add contribution guidelines

### PR DESCRIPTION
**Pull request by AlexeySachkov**
Any feedback/comments/suggestions are highly appreciated

TODO:
[ ] testing guidelines

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+For feature requests, please describe the feature you would like to see
+implemented in the translator.
+
+[new issue]: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/new
+
+## If you would like to contribute your change
+
+Please open a [pull request]. If you are not sure whether your changes are
+correct, you can either mark it as [draft] or create an issue to discuss the
+problem and possible ways to fix it prior publishing a PR.
```
Comment (svenvh) [inline CONTRIBUTING.md:21]:
```suggestion
problem and possible ways to fix it prior to publishing a PR.
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+It is okay to have several commits in the PR, but each of them should be
+build-able and tests should pass: repo maintainers can squash several commits
```
Comment (svenvh) [inline CONTRIBUTING.md:24]:
```suggestion
buildable and tests should pass. Maintainers can squash several commits
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+into a single one for you during merge, but if you would like to see several
+commits in the git history, please let us know in PR description/comments so
+maintainers will rebase your PR instead of squashing it.
+
+Each functional change (new feature or bug fix) must be supplied with
+corresponding tests. See [#testing-guidelines] for more information about
+testing. NFC (non-functional change) PRs can be accepted without new tests.
+
+Code changes should follow coding standards, which are inherited from [LLVM
+Coding Standards] - compliance of your code is checked automatically using
```
Comment (svenvh) [inline CONTRIBUTING.md:34]:
```suggestion
Coding Standards]. Compliance of your code is checked automatically using
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+Travis CI. See [clang-format] and [clang-tidy] configs for more details about
+coding standards.
+
+In order to get your PR merged, the following conditions must be met:
+- If you are first-time contributor, you have to sign
```
Comment (svenvh) [inline CONTRIBUTING.md:39]:
```suggestion
- If you are a first-time contributor, you have to sign the
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  [Contributor License Agreement]. Corresponding link and instructions will be
+  automatically posted into your PR.
+- [Travis CI testing] jobs must pass on your PR: this includes functional
+  testing and checking for complying with coding standards.
+- You need to get approval from at least one contributor with merge rights
```
Comment (svenvh) [inline CONTRIBUTING.md:44]:
```suggestion
- You need to get approval from at least one contributor with merge rights.
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+As a contributor, you should expect that even approved PR might still be left
```
Comment (svenvh) [inline CONTRIBUTING.md:46]:
```suggestion
As a contributor, you should expect that even an approved PR might still be left
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+unmerged for a few days: this is needed, because the translator is being
+developed by different vendors and individuals and we need to ensure that each
+interested party were able to react to new changes and provide feedback.
```
Comment (svenvh) [inline CONTRIBUTING.md:49]:
```suggestion
interested party is able to react to new changes and provide feedback.
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+Information below is a guideline for repo maintainers and can be used by
+contributors to get some expectations about how long PR has to be opened before
```
Comment (svenvh) [inline CONTRIBUTING.md:52]:
```suggestion
contributors to get some expectations about how long a PR has to be open before
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+it can be merged:
+- For any significant change/redesign, the PR must be opened for at least 5
```
Comment (svenvh) [inline CONTRIBUTING.md:54]:
```suggestion
- For any significant change/redesign, the PR must be open for at least 5
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  working days, so everyone interested can step in to provide feedback, discuss
+  direction and help to find bugs
```
Comment (svenvh) [inline CONTRIBUTING.md:56]:
```suggestion
  direction and help to find bugs.
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+  - Ideally, there should be more several approvals from different
```
Comment (svenvh) [inline CONTRIBUTING.md:57]:
```suggestion
  - Ideally, there should be approvals from different
```

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    vendors/individuals to get it merged
```
Comment (svenvh) [inline CONTRIBUTING.md:58]:
```suggestion
    vendors/individuals to get it merged, particularly for larger changes.
```

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+- For regular changes/bug fixes, the PR must be opened for at least 2-3 working
```
Comment (svenvh) [inline CONTRIBUTING.md:59]:
```suggestion
- For regular changes/bug fixes, the PR must be open for at least 2-3 working
```

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+  days, so everyone interested can step in for review and providing feedback
```
Comment (svenvh) [inline CONTRIBUTING.md:60]:
```suggestion
  days, so everyone interested can step in for review and provide feedback.
```

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+  - If change is vendor-specific (bug fix in vendor extension implementation or
```
Comment (svenvh) [inline CONTRIBUTING.md:61]:
```suggestion
  - If the change is vendor-specific (bug fix in vendor extension implementation or
```

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    new vendor-specific extension support), then it is okay to merge PR sooner
+  - If change affects or might affect several interesting parties, the PR must
```
Comment (svenvh) [inline CONTRIBUTING.md:63]:
```suggestion
  - If the change affects or might affect several interested parties, the PR must
```

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    be left opened for 2-3 working days and it would be good to see feedback
```
Comment (svenvh) [inline CONTRIBUTING.md:64]:
```suggestion
    be left open for 2-3 working days and it would be good to see feedback
```

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    from different vendors/inviduals before merging
+- Tiny NFC changes or trivial build fixes (due to LLVM API changes) can be
+  submitted as soon as testing is finished and PR approved - no need to wait for
+  too long
+- In general, just use common sense to wait long enough to get feedback from
+  everyone who might be interested in the PR and don't hesitate to explicitly
+  mention ones who might be interested in reviewing the PR
```
Comment (svenvh) [inline CONTRIBUTING.md:71]:
```suggestion
  mention individuals who might be interested in reviewing the PR.
```

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+- For regular changes/bug fixes, the PR must be opened for at least 2-3 working
+  days, so everyone interested can step in for review and providing feedback
+  - If change is vendor-specific (bug fix in vendor extension implementation or
+    new vendor-specific extension support), then it is okay to merge PR sooner
+  - If change affects or might affect several interesting parties, the PR must
+    be left opened for 2-3 working days and it would be good to see feedback
+    from different vendors/inviduals before merging
+- Tiny NFC changes or trivial build fixes (due to LLVM API changes) can be
+  submitted as soon as testing is finished and PR approved - no need to wait for
+  too long
```
Comment (svenvh) [inline CONTRIBUTING.md:68]:
```suggestion
  too long.
```

**Comment by Hopetech** [comment]
LGTM
I would just add some more paragraph title like `Conditions to merge a PR`.

**Comment by AlexeySachkov** [comment]
Applied all feedback from @svenvh in c908bdf0b2e00a79c36215e0fe471ce9831a81c1

> LGTM
> I would just add some more paragraph title like `Conditions to merge a PR`.

Should be done in c908bdf0b2e00a79c36215e0fe471ce9831a81c1 as well. @Hopetech please take a look and let me know if there is any feedback

**Comment by Hopetech** [comment]
Amazing, thanks.
Looks good.

**Comment by Naghasan** [comment]
LGTM, thanks

**Comment by AlexeySachkov** [comment]
I've merged current version into the repo. If you have any additional feedback or other ideas about how to improve it, please either comment here or open a PR

**Pull request merged**